File: workers/research/src/black_book_research/discovery/identity.py
"""Candidate identity and source reference helpers (BB-039)."""

from __future__ import annotations

from black_book_research.adapters.types import AdapterCandidateRecord

from .hashing import hash_candidate_content, hash_utf8
from .types import DiscoveryCandidateIdentity, SourceReference


def build_source_reference(record: AdapterCandidateRecord) -> SourceReference:
    return SourceReference(
        source_id=record.provenance.source_id,
        adapter_id=record.provenance.adapter_id,
        parser_version=record.provenance.parser_version,
        registry_entry_id=record.provenance.registry_entry_id,
        run_id=record.provenance.run_id,
        captured_at=record.provenance.captured_at,
        stable_identifier=record.stable_identifier,
        source_item_id=record.provenance.source_item_id,
    )


def candidate_identity_key(stable_identifier: str, content_hash_digest: str) -> str:
    material = f"{stable_identifier.strip()}::sha256:{content_hash_digest}"
    return hash_utf8(material).digest[:32]


def build_candidate_identity(record: AdapterCandidateRecord) -> DiscoveryCandidateIdentity:
    content_hash = hash_candidate_content(record)
    return DiscoveryCandidateIdentity(
        identity_key=candidate_identity_key(record.stable_identifier, content_hash.digest),
        stable_identifier=record.stable_identifier,
        content_hash=content_hash,
        source_references=(build_source_reference(record),),
    )

File: packages/domain/src/discovery/identity.ts
/**
 * Candidate identity and source reference helpers.
 */
import type { AdapterCandidateRecord } from '../adapters/types.js';
import { hashUtf8, type ContentHash } from '../provenance/hashes.js';
import { hashCandidateContent } from './hashing.js';
import type { DiscoveryCandidateIdentity, SourceReference } from './types.js';

export function buildSourceReference(record: AdapterCandidateRecord): SourceReference {
  return {
    sourceId: record.provenance.sourceId,
    adapterId: record.provenance.adapterId,
    parserVersion: record.provenance.parserVersion,
    registryEntryId: record.provenance.registryEntryId,
    runId: record.provenance.runId,
    capturedAt: record.provenance.capturedAt,
    ...(record.provenance.sourceItemId !== undefined
      ? { sourceItemId: record.provenance.sourceItemId }
      : {}),
    stableIdentifier: record.stableIdentifier,
  };
}

export function candidateIdentityKey(stableIdentifier: string, contentHash: ContentHash): string {
  const material = `${stableIdentifier.trim()}::${contentHash.algorithm}:${contentHash.digest}`;
  return hashUtf8(material).digest.slice(0, 32);
}

export function buildCandidateIdentity(record: AdapterCandidateRecord): DiscoveryCandidateIdentity {
  const contentHash = hashCandidateContent(record);
  return {
    identityKey: candidateIdentityKey(record.stableIdentifier, contentHash),
    stableIdentifier: record.stableIdentifier,
    contentHash,
    sourceReferences: [buildSourceReference(record)],
  };
}
